[#598] Expose native audio events

LoserAntbear opened the pull request:
Closes #598 

## Changes:

- Add `AudioContext` And `AudioContextOperator` Singleton for handling multiple instances of `AudioContext`.
- Add `AudioInstance` with `WebAudioInstance` and `AudioTagInstance` implementations.
- Add `MediaEvent` which extends `GameEvent` with `NativeSoundEvent` implementation.
- `Sound` now extends `Resource` with all functionality. `Sound` now could be signed on:
  - 'volumechange' - on playing sound volume change;
  - 'pause' - on playback pause;
  - 'stop - on playback stop';
  - 'emptied' - on data cleanup;
  - 'resume' - on playback resume;
  - 'playbackstart' - on playback start;
  - 'playbackend' - on playback end;
- Move `WebAudio` to `Utils` accordingly to it's purpose.

Working on tests and `CHANGELOG` entries.

LoserAntbear commented:
> expose native sudio dound events through excalibur 
That was rough night >.<

LoserAntbear commented on src/engine/Resources/Sound/AudioContext.ts:
```diff
@@ -0,0 +1,100 @@
+      return (id > -1) ? id : null;
+   }
+
+   // private _getAudioCtxtByIndex(index: number): AudioContext {
```
Sorry, i sought it'd be useful to create this methods in advance, looks like they'll be needed some day. But okay, i'll remove them.

LoserAntbear commented on src/engine/Util/Util.ts:
```diff
@@ -1,7 +1,14 @@
+/**
+ * Export Sound
+ */
+export {
```
Yeah, i meant to export `canPlayFile` from `./Util/Sound`. I've changed comment to better understandability.

eonarheim commented on src/engine/Resources/Sound/AudioContext.ts:
```diff
@@ -0,0 +1,100 @@
+export class AudioContextFactory {
+   public static create() {
+      if ((<any>window).AudioContext || (<any>window).webkitAudioContext) {
```
I'm fairly certain we want only 1, we tried using multiple `AudioContext`s in the past. At the time browsers did not handle that well. 

A single `AudioContext` can play multiple concurrent audio streams, so there is no reason to support multiple.

eonarheim replied:
What would a `OfflineAudioContext` do?

LoserAntbear replied:
Typically  `OfflineAudioContext` is used to pre-process audio data and than pass it to `AudioBuffer`. One of use-cases is described here:
When you want to create real-time canvas visualization (https://stackoverflow.com/questions/8074152/is-there-a-way-to-use-the-web-audio-api-to-sample-audio-faster-than-real-time).

and about multiple `AudioContexts`, maybe, we could create a pending issue, waiting for better support in browsers? Because `AudioContext` has useful methods like `suspend`, `resume`, etc. it could be useful while creating sound collections. Particularly, we could use multiple `AudioContext`s for issue #590 . Dividing sound in groups like "music", "ambient sounds", "sfx" etc. Handling each group in bounds of separate `AudioContext` would be pretty neat, i think.

eonarheim commented on src/engine/Events/GameEvent.ts:
```diff
@@ -1,85 +1,13 @@
-export type pointerdragenter = 'pointerdragenter';
-export type pointerdragleave = 'pointerdragleave';
-export type pointerdragmove = 'pointerdragmove';
+import { Scene } from '../Scene';
```
Let's not refactor `Events.ts` at this time. It causes a lot of changes in files unrelated to audio.

eonarheim commented on src/engine/UIActor.ts:
```diff
@@ -21,7 +21,7 @@ export class UIActor extends Actor {
     */
    constructor(xOrConfig?: number | IActorArgs, y?: number, width?: number, height?: number) {
-      if (typeof(xOrConfig !== 'object')) {
+      if (xOrConfig instanceof Object) {
```
Let's revert changes to `UIActor`, this is unrelated to Audio

LoserAntbear commented:
So. Force-updated this one.
- Undid all changes related to moving of `Events.ts` to new namespace.
- Removed `AudioContextOperator` in favor of retuniing only single instance of `AudioContext` by `AudioContextFactory`
- And reverted back all Unrelated changes.

LoserAntbear commented:
Fixed `CHANGELOG`  respectively.

eonarheim commented:
@LoserAntbear Thanks again!

LoserAntbear commented:
@eonarheim trying my best. Thank you, guys, too, btw, for letting me help you!
